### ISSUE TITLE
Unity Editor (2021.1.10f) hangs in Application.Reload if Connection.SendMessage(IntPtr, int, SendType) was used in previous test

### ISSUE THREAD

**Issue by andefoo** [Facepunch/Facepunch.Steamworks#585]
**Describe the bug**

If Connection.SendMessage(IntPtr, int, SendType) is used while testing inside Unity Editor (2021.1.15f), testing is stopped and restarted, Unity Editor hangs while doing Application.Reload and needs to killed. This may be a Unity bug but nevertheless, it is very annoying.

By binary searching, I noticed it is caused by this commit: https://github.com/Facepunch/Facepunch.Steamworks/commit/5c11508f41a638eb27398234c29ec1cdf051465d#diff-84f9225b3c0fa0331b4775665789c00dd043e02674b5f83e02fd5af6d7b810d0
 
Same hang can happen for various reasons it seems, here is a related thread: https://forum.unity.com/threads/application-reload.989720/

I got rid of the hang by replacing the new version of Connection.SendMessage(IntPtr, int, SendType) by the one it was before that commit. I don't know what other problems that might cause though. But could there be some other issues like leaks with that or the new broadcast code?

**To Reproduce**
1. Take the latest (or anything after this submit: https://github.com/Facepunch/Facepunch.Steamworks/commit/5c11508f41a638eb27398234c29ec1cdf051465d#diff-84f9225b3c0fa0331b4775665789c00dd043e02674b5f83e02fd5af6d7b810d0) version of the library, compile it and put it to your Unity project.
2. Test a project that uses Connection.SendMessage(IntPtr, int, SendType) inside Unity Editor (2021.1.10f)
3. Stop testing
4. Start testing again
5. Observe that Unity gets stuck in Application.Reload forever. 

**Expected behavior**
Unity should not hang but a new test should start. 

**Desktop (please complete the following information):**
 - OS: Windows 10
- Unity 2021.1.15f

**Comment by Rohansi** [comment]
Are you shutting down Steamworks when the game/playmode exits?

**Comment by andefoo** [comment]
I was going to say yes but then I realized there is a bug in my shutdown code and it was not called. Unfortunately, I can't right now test if Shutdown fixes the issue, too (I can test that later if you need the info). But test sessions often end ungracefully anyway, for example by pressing Stop to save time, so it would still be an issue even if it did fix the issue.

**Comment by Rohansi** [comment]
In Rust we shut it down in an `OnDestroy` and it's always called - even when abruptly ending play mode within the editor. We have gotten Unity hangs with or without this due to not shutting down Steamworks so it'd be good to confirm if that fixes it.

**Comment by andefoo** [comment]
Yes, calling Shutdown does help and the Application.Reload hang does not happen if it was called at the end of the previous test session. As this could very well be a Unity bug anyway, I am not sure if there is anything you should do about it.

I will probably keep my backed out version at least for now in my own tests since with that, the hang does not happen even if I end testing without Shutdown. But what exactly was the point of that commit? How does it make sending messages better? Is the performance better and/or are there less memory allocations or something else?

**Comment by Rohansi** [comment]
> I will probably keep my backed out version at least for now in my own tests since with that, the hang does not happen even if I end testing without Shutdown. But what exactly was the point of that commit? How does it make sending messages better? Is the performance better and/or are there less memory allocations or something else?

Yes, it improves performance by reducing allocations within Steam Networking itself.

It's looking a lot like the same Unity bug you sent. The `BufferManager` class has a `free` function which could be called on any Steamworks thread so it will be waiting until that thread shuts down. Either Unity needs to fix that or that issue needs to be worked around somehow... but I can't think of any possible way to workaround that Unity behavior. Shutting things down works.

Even in your branch you could still hit this issue if you don't shutdown and use `SendMessages` to send the same message to multiple connections. This can be significantly faster than sending all the messages separately since it's one pooled allocation vs. `n` allocations.
